### PR TITLE
SymPyExpression

### PR DESCRIPTION
**Pull request by hsiang-wu**
Implemented `SymPyExpression` and `SymPyInterpreter`'s `eval()` method.

### CODE REVIEW

**Review by rodrigodesalvobraz — CHANGES_REQUESTED** [review]
Very nice!

**Comment by hsiang-wu** [comment]
the above commits are me merging my forked branch to origin/main

File: neuralpp/symbolic/sympy_expression.py [hunk 1]
```diff
@@ -0,0 +1,160 @@
+from __future__ import annotations
+from typing import List, Any, Type, Callable, Tuple
+from abc import ABC
+
+import sympy
+import operator
+import builtins
+from neuralpp.symbolic.expression import Expression, FunctionApplication, Variable, Constant
+
+# In this file's doc, I try to avoid the term `sympy expression` because it could mean both sympy.Expr/Basic
```
[the remark this answers is not in the document]
Comment (hsiang-wu) [reply neuralpp/symbolic/sympy_expression.py:10]: done

File: neuralpp/symbolic/sympy_interpreter.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+from neuralpp.symbolic.expression import Expression, FunctionApplication, Constant, Variable
+from neuralpp.symbolic.simplifier import Simplifier
+from neuralpp.symbolic.interpreter import Interpreter
+from neuralpp.symbolic.sympy_expression import SymPyExpression, is_sympy_value
+from typing import Dict
+
+
+def context_to_variable_value_map_helper(context: Expression,
+                                         variable_to_value: Dict[sympy.Symbol, int | sympy.Integer]) \
+        -> Dict[sympy.Symbol, int | sympy.Integer]:
```
[the remark this answers is not in the document]
Comment (hsiang-wu) [reply neuralpp/symbolic/sympy_interpreter.py:12]: yes

File: neuralpp/symbolic/sympy_interpreter.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
+import sympy
+
+from neuralpp.symbolic.expression import Expression, FunctionApplication, Constant, Variable
+from neuralpp.symbolic.simplifier import Simplifier
+from neuralpp.symbolic.interpreter import Interpreter
+from neuralpp.symbolic.sympy_expression import SymPyExpression, is_sympy_value
+from typing import Dict
+
+
+def context_to_variable_value_map_helper(context: Expression,
```
[the remark this answers is not in the document]
Comment (hsiang-wu) [reply neuralpp/symbolic/sympy_interpreter.py:10]: I don't think we need the plural form here? Would "dict" suffice?

**Comment by hsiang-wu** [comment]
Fixes

File: neuralpp/symbolic/sympy_interpreter.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (rodrigodesalvobraz) [reply neuralpp/symbolic/sympy_interpreter.py]: Sure, that's what I had in mind.

**Review by CactusWin — COMMENTED** [review]
Nice work!

File: neuralpp/test/quick_tests/symoblic/expression_test.py [hunk 4]
```diff
@@ -84,3 +111,43 @@ def test_basic_function_application():
...
+    assert fa.subexpressions[2].value == 2
+    assert fa == fa
+
+    assert fa != SymPyExpression.new_function_application(add_func, [constant_one, constant_one])
+    assert fa == SymPyExpression.new_function_application(add_func, [constant_one, constant_two])
+
+    assert constant_one == SymPyConstant(sympy.Integer(1))
+
+    # replace() is also deep and not in-place (returns a new object instead of modify the called on)
+    fa4 = SymPyExpression.new_function_application(add_func, [constant_one, fa])
```
[the remark this answers is not in the document]
Comment (hsiang-wu) [reply neuralpp/test/quick_tests/symoblic/expression_test.py:140]: no...it's resulted from copy-paste of code and deletion of `fa2` and `fa3`

**Review by rodrigodesalvobraz — APPROVED** [review]

**Pull request merged**